### PR TITLE
Disable tracking 404

### LINKED ISSUE #7565: Random 404s from scanning bots create contacts only in segment with tag exclusion filter
## Bug Description

| Q   | A
| --- | ---
| Mautic version | 2.15.1
| PHP version | 7.1.27
| Browser | Chrome 71.0.3578.98

### Steps to reproduce

1. Create fresh installation which is directly accessible from the internet
1. Create a couple of tags
1. Create a segment with a filter which excludes those tags
1. Wait a few hours for bots which scan for security vulnerabilities to find the IP and start their attacks, filling the logs with 404s
1. Wait for the next `mautic:segments:update` cron job to run

### Observed results

The contacts get added to the segment by the cron job, e.g.:

        mautic_1  | [CRON] Rebuilding contacts for segment 2
        mautic_1  | [CRON] 213 total contact(s) to be added in batches of 300
        mautic_1  | [CRON] 0/213 [>---------------------------]   0%
        mautic_1  | [CRON] 21/213 [==>-------------------------]   9%
        mautic_1  | [CRON] 42/213 [=====>----------------------]  19%
        mautic_1  | [CRON] 63/213 [========>-------------------]  29%
        mautic_1  | [CRON] 84/213 [===========>----------------]  39%
        mautic_1  | [CRON] 105/213 [=============>--------------]  49%
        mautic_1  | [CRON] 126/213 [================>-----------]  59%
        mautic_1  | [CRON] 147/213 [===================>--------]  69%
        mautic_1  | [CRON] 168/213 [======================>-----]  78%
        mautic_1  | [CRON] 189/213 [========================>---]  88%
        mautic_1  | [CRON] 210/213 [===========================>]  98%
        mautic_1  | [CRON] 213/213 [============================] 100%
        mautic_1  | [CRON] 0 total contact(s) to be removed in batches of 300
        mautic_1  | [CRON] 213 contact(s) affected

After this happened a few times, visiting https://my.mautic.site/s/segments shows `View 286 Contacts` in the `# contacts` column.

This is the first issue, although I believe this was already reported in 2015 in issue #1060.  However there is an additional problem:

When clicking on that to view contacts using a `segment:my-segment-alias` filter, it shows:

    No Results Found

    Seems there are none! Try changing a filter (if applicable) or how about creating a new one?

Similarly, when visiting https://my.mautic.site/s/contacts without a filter, it only shows valid contacts, and none of these spammer contacts.

Despite this, individual contacts can be seen by clicking on the segment view, which goes to a URL like https://my.mautic.site/s/segments/view/1, and then they are all listed at the bottom of the page under the `Contacts` tab.

### Expected results

404 hits from previously unrecognised sources do not add any contacts to the database, because this is just inevitable background noise suffered by every website out there on the internet.

In contrast, there *is* value in tracking 404 hits from previously *recognised* sources, since this could yield valuable information about legitimate, well-intentioned visitors hitting broken links.

### Initial debugging

I enabled the MySQL query log:

    set global general_log=on;

and then visited https://my.mautic.site/s/contacts to reverse-engineer how Mautic compiles the list of contacts to display.  AFAICS, it's using this query:

    SELECT l.* FROM leads l LEFT JOIN users u ON u.id = l.owner_id WHERE l.date_identified IS NOT NULL ORDER BY l.last_active DESC LIMIT 30 OFFSET 0

and indeed the `date_identified` field being `NULL` is what distinguishes the genuine contacts from the spam ones which are polluting the database:

    mysql> select isnull(date_identified), count(isnull(date_identified)) from leads group by isnull(date_identified);
    +-------------------------+--------------------------------+
    | isnull(date_identified) | count(isnull(date_identified)) |
    +-------------------------+--------------------------------+
    |                       0 |                              1 |
    |                       1 |                            288 |
    +-------------------------+--------------------------------+
    2 rows in set (0.00 sec)

How can we progress this to a fix?

It's worth noting that only segments which have this tag exclusion filter get polluted.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/74711390-random-404s-from-scanning-bots-create-contacts-only-in-segment-with-tag-exclusion-filter?utm_campaign=plugin&utm_content=tracker%2F5355074&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F5355074&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by kuzmany**
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | fixes https://github.com/mautic/mautic/issues/7565
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Based on discussion https://github.com/mautic/mautic/issues/7565 and comment https://github.com/mautic/mautic/issues/7565#issuecomment-497675707 
we've added option Do Not Track 404 error page tracking for anonymous visitors. This causes a lot of anonymous contacts from bots. New option added to Configuration > Tracking Settings tab.  
UNIT test included in this PR

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Try in incognito not existed url: https://yourmautic.com/whatever
2. Go to Mautic > Contacts and switch to anonymous list view
3. See new contact was create with https://yourmautic.com/whatever page hit

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Repeat all steps but disable in Configuration > Tracking Settings tab > Do Not Track 404 error page
3. See If 404 not found page wasn't tracked
4. Create identified contact (via form, via tracking code)
5. Go to 404 not found page and see If page was tracked

image

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/7614"></a>

### CODE REVIEW

**Review by indorock — DISMISSED** [review]

**Review by aspiers — COMMENTED** [review]
Hi, thanks a lot for doing this!!  It looks great but it's not clear to me whether the option disables tracking of *all* 404s, or just anonymous 404s?  As explained in https://github.com/mautic/mautic/issues/7565#issuecomment-497675707 it is important that 404s generated by known contacts are still tracked, even when the option is enabled.

File: app/bundles/PageBundle/Translations/en_US/messages.ini [hunk 1]
```diff
@@ -33,6 +33,8 @@ mautic.page.config.form.google.analytics="Analytics script (i.e. Google Analytic
 mautic.page.config.form.google.analytics.tooltip="Insert the analytics script to have it automatically included in the soure of landing pages."
 mautic.page.config.form.track_contact_by_ip="Identify visitors by IP"
 mautic.page.config.form.track_contact_by_ip.tooltip="If enabled, unidentified visitors with the same IP address as an existing contact will be tracked as that contact."
+mautic.page.config.form.disable_tracking_404="Disable tracking 404 error page"
+mautic.page.config.form.disable_tracking_404.tooltip="If enabled, not founded page will not tracking"
```
Comment (ghost) [inline app/bundles/PageBundle/Translations/en_US/messages.ini]:
```suggestion
mautic.page.config.form.disable_tracking_404.tooltip="If enabled, unidentified visits to 404 response code error pages will not be tracked."
```

File: app/bundles/PageBundle/Translations/en_US/messages.ini [hunk 2]
```diff
@@ -33,6 +33,8 @@ mautic.page.config.form.google.analytics="Analytics script (i.e. Google Analytic
 mautic.page.config.form.google.analytics.tooltip="Insert the analytics script to have it automatically included in the soure of landing pages."
 mautic.page.config.form.track_contact_by_ip="Identify visitors by IP"
 mautic.page.config.form.track_contact_by_ip.tooltip="If enabled, unidentified visitors with the same IP address as an existing contact will be tracked as that contact."
+mautic.page.config.form.disable_tracking_404="Disable tracking 404 error page"
```
Comment (ghost) [inline app/bundles/PageBundle/Translations/en_US/messages.ini]:
```suggestion
mautic.page.config.form.disable_tracking_404="Disable 404 error page tracking"
```

**Review by ghost — CHANGES_REQUESTED** [review]
Requesting changes to en_US entries for improved grammar, punctuation.

**Comment by kuzmany** [comment]
@mightyfree thank you

@aspiers at the moment, client did pay just for this feature. 
I agree with your suggestion, but there is no budget for extra hours. If you are interest let me now and I can extend it.

**Comment by indorock** [comment]
> @mightyfree thank you
> 
> @aspiers at the moment, client did pay just for this feature.
> I agree with your suggestion, but there is no budget for extra hours. If you are interest let me now and I can extend it.

@kuzmany how much more time do you estimate you'll need to add that extra change?

**Comment by kuzmany** [comment]
thanks to @indorock  donate
I've added support for disable tracking 404 page for anonymous contacts
PR, description and steps to reproduce/tests was updated.

Please re-test @aspiers  @indorock

**Comment by RCheesley** [comment]
Tried to test this in Mautibox but it says "Patch could not be applied. Try rebasing the branch for this pull request."

File: app/bundles/PageBundle/Translations/en_US/messages.ini [hunk 3]
```diff
@@ -33,6 +33,8 @@ mautic.page.config.form.google.analytics="Analytics script (i.e. Google Analytic
 mautic.page.config.form.google.analytics.tooltip="Insert the analytics script to have it automatically included in the soure of landing pages."
```
Comment (virgilwashere) [inline app/bundles/PageBundle/Translations/en_US/messages.ini]: This was updated in 5f9139c, and fails against [release:2.15.3](/mautic/mautic/tree/2.15.3).

```suggestion
mautic.page.config.form.google.analytics.tooltip="Insert the analytics script to have it automatically included in the source of landing pages."
```

Has been working well on 2.15.1/2.15.2

**Review by virgilwashere — CHANGES_REQUESTED** [review]
Has been working well on 2.15.1/2.15.2.

File: app/bundles/PageBundle/Translations/en_US/messages.ini (hunk 3 above)
Comment (virgilwashere) [reply app/bundles/PageBundle/Translations/en_US/messages.ini]: I think the PR still needs a rebase.

[Mautibox agrees](https://mautibox.com/7614):
> Patch could not be applied. Try rebasing the branch for this pull request.

**Review by virgilwashere — APPROVED** [review]
(pressed the wrong button)

**Review by virgilwashere — CHANGES_REQUESTED** [review]
### rebase required

I think the PR still needs a rebase.

[Mautibox agrees](https://mautibox.com/7614):
> Patch could not be applied. Try rebasing the branch for this pull request.


### applying the current version

However, I successfully applied the PR using

```console
$ git apply --verbose --recount -C1 7614.diff
```

which just checks [lines 35-38](https://github.com/mautic/mautic/blob/0bc4e7e9e0ccec9a58ecdcad798e1c951049fceb/app/bundles/PageBundle/Translations/en_US/messages.ini#L35-L38) of `messages.ini`

https://github.com/mautic/mautic/blob/0bc4e7e9e0ccec9a58ecdcad798e1c951049fceb/app/bundles/PageBundle/Translations/en_US/messages.ini#L35-L38

**Comment by kuzmany** [comment]
@virgilwashere ready to test https://mautibox.com/7614

**Comment by virgilwashere** [comment]
@kuzmany I had a problem where contacts were not being created properly from a form submission. 

Form has firstname last name and email fields set to save to contact. But result is a contact with IP address only. 
Form fields are listed in History.

But if I toggle 'Skip 404' to 'No' the form submission saves the contact correctly.

**Comment by npracht** [comment]
Hi there!
It has been decided to not create any extra Mautic 2 minor versions (which means no more features in M2) knowing that Mautic 3 is coming very soon.

**We now want to integrate your contribution in the Mautic 3 roadmap as 3.1.0 candidate.**

## How to do?
1. Check if your feature or enhancement is still missing and relevant in Mautic 3
2. **If Yes**: rebase your PR against the 3.x branch
3. **If No**: please consider closing your PR if you are absolutely sure that the feature has made it into Mautic 3 already

**Please report results by commenting on your PR to make us administration easier.**
______
You can more information on how to do all of that on [this blog post "Getting you PR ready for Mautic 3"](https://www.mautic.org/blog/developer/getting-your-pr-ready-mautic-3).

**Comment by RCheesley** [comment]
@kuzmany another with conflicts to be resolved :)

**Comment by RCheesley** [comment]
Bumping to 3.2 due to unresolved conflicts

**Comment by RCheesley** [comment]
@cla-bot check

**Comment by kuzmany** [comment]
Conflicts resolved

**Comment by andres-asm** [comment]
let's hope this is merged soon, this is still an issue and I hope it can be fixed in m4

**Review by jos0405 — DISMISSED** [review]
This works as expected. After turning on the feature any 404 on the Mautic (sub)domain is not tracked. Very much needed feature.

**Review by mabumusa1 — DISMISSED** [review]
I checked this PR and implemented on a live server with tracking on. I sent a request via Paython bot to 404 page, the contact was not marked.

I also tried from a browser, a contact was not created.

I think this PR is read to go once we get phpstan working

**Review by mabumusa1 — APPROVED** [review]
Thanks @kuzmany, after the new changes I think this PR is ready to commit

**Comment by tomekkowalczyk** [comment]
@kuzmany I understand that the mechanism works only on Mautic websites? If I place the tracking code on an external website and visit the 404 website, an anonymous contact is still created.

**Comment by kuzmany** [comment]
@tomekkowalczyk it cover all 404 produced by Mautic, not from website

**Pull request merged**